platform/linux-generic/odp_ipsec.c

```diff
@@ -565,18 +593,40 @@ static int ipsec_in_esp(odp_packet_t *pkt,
 	param->auth_iv_ptr = state->iv;
 
 	state->esp.aad.spi = esp.spi;
-	state->esp.aad.seq_no = esp.seq_no;
 	state->in.seq_no = odp_be_to_cpu_32(esp.seq_no);
+	if (ipsec_sa->esn)
+		ipsec_sa->in.seq64 = ipsec_compute_esn(ipsec_sa, state->in.seq_no);
+
+	/* AAD with ESN */
+	if (ipsec_sa->esn && (!ipsec_sa->seqh_len))
+		state->esp.aad.seq_no64 = odp_cpu_to_be_64(ipsec_sa->in.seq64);
+	else if (!ipsec_sa->esn)
+		state->esp.aad.seq_no = esp.seq_no;
```

Comment:
To me this construct is hard to read. Instead of this:
```
if (a & !b)
        x();
else if (!a)
        y();
```
...one could write this:
```
if (a) {
        if (!b)
                x();
} else
        y();
```

Reply:
ack, will make it readable.